File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/AbstractTask.java
package com.roxiemobile.networkingapi.network.rest.request;

import com.roxiemobile.androidcommons.concurrent.ThreadUtils;
import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.networkingapi.network.http.HttpHeaders;
import com.roxiemobile.networkingapi.network.http.HttpStatus;
import com.roxiemobile.networkingapi.network.rest.CallResult;
import com.roxiemobile.networkingapi.network.rest.Callback;
import com.roxiemobile.networkingapi.network.rest.Cancellable;
import com.roxiemobile.networkingapi.network.rest.HttpBody;
import com.roxiemobile.networkingapi.network.rest.HttpResult;
import com.roxiemobile.networkingapi.network.rest.RestApiClient;
import com.roxiemobile.networkingapi.network.rest.Task;
import com.roxiemobile.networkingapi.network.rest.TaskQueue;
import com.roxiemobile.networkingapi.network.rest.config.DefaultHttpClientConfig;
import com.roxiemobile.networkingapi.network.rest.config.HttpClientConfig;
import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;
import com.roxiemobile.networkingapi.network.rest.response.RestApiError;
import com.roxiemobile.networkingapi.network.rest.response.error.ApplicationLayerError;
import com.roxiemobile.networkingapi.network.rest.response.error.TransportLayerError;
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ConnectionException;
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ResponseException;
import com.roxiemobile.networkingapi.network.rest.routing.HttpRoute;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;

public abstract class AbstractTask<Ti extends HttpBody, To>
        implements Task<Ti, To>, Cancellable {

// MARK: - Construction

    protected AbstractTask(@NotNull TaskBuilder<Ti, To> builder) {
        Guard.notNull(builder, "builder is null");

        // Init instance variables
        mTag = builder.tag();
        mRequestEntity = builder.requestEntity();
    }

// MARK: - Properties

    /**
     * The tag associated with a task.
     */
    public final @Nullable String tag() {
        return mTag;
    }

    /**
     * The original request entity.
     */
    public final @NotNull RequestEntity<Ti> requestEntity() {
        return mRequestEntity;
    }

// MARK: - Methods

    /**
     * Synchronously send the request and return its response.
     */
    @Override
    public final void execute(@Nullable Callback<Ti, To> callback) {
        boolean shouldExecute = true;

        CallResult<To> result = null;
        try {

            // Check if task must be executed
            if (callback != null) {
                shouldExecute = callback.onShouldExecute(this);
            }

            // Execute task if needed
            if (shouldExecute) {
                result = call();
            }
        }
        catch (Throwable ex) {
            result = CallResult.failure(new ApplicationLayerError(ex));
        }

        // Yielding result to listener
        if (callback != null && shouldExecute) {
            yield(result, callback);
        }
    }

    /**
     * TODO
     */
    @Override
    public final @NotNull Cancellable enqueue(@Nullable Callback<Ti, To> callback, boolean callbackOnUiThread) {
        return TaskQueue.enqueue(this, callback, callbackOnUiThread);
    }

    /**
     * Performs the request and returns the response, or throws an exception if unable to do so.
     * May return null if this call was canceled.
     */
    protected final @Nullable CallResult<To> call() throws Exception {
        Guard.isFalse(ThreadUtils.runningOnUiThread(), "This method must not be called from the main thread!");
        CallResult<To> result = null;

        // Send request to the server
        HttpResult httpResult = callExecute();
        RestApiError error = null;

        // Are HTTP response is still needed?
        if (!isCancelled()) {

            // Handle HTTP response
            if (httpResult.isSuccess()) {

                ResponseEntity<byte[]> entity = httpResult.value();
                HttpStatus status = entity.status();

                // Create a new call result
                if (status.is2xxSuccessful()) {
                    result = onSuccess(CallResult.success(entity));
                }
                else {
                    ResponseException cause = new ResponseException(entity);
                    // Build application layer error
                    error = new ApplicationLayerError(cause);
                }
            }
            else {
                Throwable cause = httpResult.error();

                // Wrap up HTTP connection error
                if (cause instanceof IOException) {
                    cause = new ConnectionException(cause);
                }

                // Build transport layer error
                error = new TransportLayerError(cause);
            }

            // Handle error
            if (error != null) {
                result = onFailure(error);
            }
        }
        else {

            // Handle request cancellation
            onCancel();
        }

        // Done
        return result;
    }

    protected abstract @NotNull HttpResult callExecute();

    /**
     * TODO
     */
    protected final @NotNull RestApiClient createClient() {

        // Get HTTP client config
        HttpClientConfig httpClientConfig = httpClientConfig();

        // Create/init HTTP client
        RestApiClient.Builder builder = new RestApiClient.Builder()
                .httpClientConfig(httpClientConfig);

        // Done
        return builder.build();
    }

    /**
     * TODO
     */
    protected @NotNull HttpClientConfig httpClientConfig() {
        return DEFAULT_HTTP_CLIENT_CONFIG;
    }

    /**
     * TODO
     */
    protected @NotNull RequestEntity<HttpBody> createRequestEntity(@NotNull HttpRoute route) {
        // Create HTTP request entity
        return new BasicRequestEntity.Builder<>(requestEntity(), httpBody())
                .uri(route.toURI())
                .headers(httpHeaders())
                .build();
    }

    /**
     * TODO
     */
    protected @NotNull HttpHeaders httpHeaders() {
        return HttpHeaders.readOnlyHttpHeaders(requestEntity().headers());
    }

    /**
     * TODO
     */
    protected @Nullable HttpBody httpBody() {
        return requestEntity().body();
    }

    /**
     * TODO
     */
    protected abstract @NotNull CallResult<To> onSuccess(@NotNull CallResult<byte[]> httpResult);

    /**
     * TODO
     */
    protected @NotNull CallResult<To> onFailure(@NotNull RestApiError error) {
        Guard.notNull(error, "error is null");
        return CallResult.failure(error);
    }

    /**
     * TODO
     */
    protected void onCancel() {
        // Do nothing
    }

    /**
     * TODO
     */
    @Override
    public final @NotNull Task<Ti, To> clone() {
        return createBuilder().build();
    }

    /**
     * TODO
     */
    protected abstract @NotNull TaskBuilder<Ti, To> createBuilder();

    /**
     * TODO
     */
    public final boolean cancel() {
        return !mCancelled.getAndSet(true);
    }

    /**
     * TODO
     */
    public final boolean isCancelled() {
        return mCancelled.get();
    }

// MARK: - Private Methods

    private void yield(@Nullable CallResult<To> result, @NotNull Callback<Ti, To> callback) {
        Guard.notNull(callback, "callback is null");

        if (isCancelled()) {
            callback.onCancel(this);
        }
        else {
            if (result != null) {
                if (result.isSuccess()) {
                    callback.onSuccess(this, result.value());
                }
                else {
                    callback.onFailure(this, result.error());
                }
            }
            else {
                throw new IllegalStateException("!isCancelled() && (result == null)");
            }
        }
    }

// MARK: - Inner Types

    public abstract static class Builder<Ti, To, BuilderType extends TaskBuilder<Ti, To>>
            implements TaskBuilder<Ti, To> {

        public Builder() {
            // Do nothing
        }

        protected Builder(@NotNull Task<Ti, To> task) {
            // Init instance variables
            mTag = task.tag();
            mRequestEntity = task.requestEntity();
        }

        public @Nullable String tag() {
            return mTag;
        }

        public @NotNull BuilderType tag(@Nullable String tag) {
            mTag = tag;
            //noinspection unchecked
            return (BuilderType) this;
        }

        public @Nullable RequestEntity<Ti> requestEntity() {
            return mRequestEntity;
        }

        public @NotNull BuilderType requestEntity(@NotNull RequestEntity<Ti> request) {
            mRequestEntity = request;
            //noinspection unchecked
            return (BuilderType) this;
        }

        public @NotNull Task<Ti, To> build() {
            checkInvalidState();
            return createTask();
        }

        protected void checkInvalidState() {
            Guard.notNull(mRequestEntity, "requestEntity is null");
            Guard.notNull(mRequestEntity.uri(), "requestEntity.uri is null");
        }

        protected abstract @NotNull Task<Ti, To> createTask();

        private @Nullable String mTag;
        private @Nullable RequestEntity<Ti> mRequestEntity;
    }

// MARK: - Constants

    private static final @NotNull HttpClientConfig DEFAULT_HTTP_CLIENT_CONFIG =
            new DefaultHttpClientConfig();

// MARK: - Variables

    private final @Nullable String mTag;

    private final @NotNull RequestEntity<Ti> mRequestEntity;

    private final @NotNull AtomicBoolean mCancelled = new AtomicBoolean(false);
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/AbstractTask.kt
@file:Suppress("DEPRECATION", "UNCHECKED_CAST", "unused")

package com.roxiemobile.networkingapi.network.rest.request

import com.roxiemobile.androidcommons.concurrent.ThreadUtils
import com.roxiemobile.java.util.toBase62
import com.roxiemobile.networkingapi.network.http.CookieStore
import com.roxiemobile.networkingapi.network.http.HttpHeaders
import com.roxiemobile.networkingapi.network.http.body.HttpBody
import com.roxiemobile.networkingapi.network.rest.CallResult
import com.roxiemobile.networkingapi.network.rest.Callback
import com.roxiemobile.networkingapi.network.rest.Cancellable
import com.roxiemobile.networkingapi.network.rest.HttpResult
import com.roxiemobile.networkingapi.network.rest.RestApiClient
import com.roxiemobile.networkingapi.network.rest.Task
import com.roxiemobile.networkingapi.network.rest.TaskQueue
import com.roxiemobile.networkingapi.network.rest.configuration.DefaultHttpClientConfig
import com.roxiemobile.networkingapi.network.rest.configuration.HttpClientConfig
import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity
import com.roxiemobile.networkingapi.network.rest.response.RestApiError
import com.roxiemobile.networkingapi.network.rest.response.error.ApplicationLayerError
import com.roxiemobile.networkingapi.network.rest.response.error.TransportLayerError
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ConnectionException
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ResponseException
import java.io.IOException
import java.net.URI
import java.util.UUID
import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean

abstract class AbstractTask<Ti: HttpBody, To>:
    Task<Ti, To>,
    Cancellable {

// MARK: - Construction

    protected constructor(builder: Builder<Ti, To, *>) {
        this.tag = builder.tag ?: createTag()
        this.requestEntity = requireNotNull(builder.requestEntity) { "requestEntity is null" }
        this.httpClientConfig = builder.httpClientConfig ?: DefaultHttpClientConfig.SHARED
    }

// MARK: - Properties

    /**
     * The tag associated with a task.
     */
    final override val tag: String

    /**
     * The original request entity.
     */
    final override val requestEntity: RequestEntity<Ti>

    /**
     * TODO
     */
    val httpClientConfig: HttpClientConfig

// MARK: - Methods

    /**
     * Synchronously send the request and return its response.
     */
    final override fun execute(callback: Callback<Ti, To>?) {
        var shouldExecute = true

        var callResult: CallResult<To>? = null
        try {

            // Check if task must be executed
            if (callback != null) {
                shouldExecute = callback.onShouldExecute(this)
            }

            // Execute task if needed
            if (shouldExecute) {
                callResult = call()
            }
        }
        catch (ex: Throwable) {
            callResult = CallResult.failure(ApplicationLayerError(ex))
        }

        // Yielding result to listener
        if (callback != null && shouldExecute) {
            yieldTo(callback, callResult)
        }
    }

    /**
     * TODO
     */
    final override fun enqueue(callback: Callback<Ti, To>?, callbackOnUiThread: Boolean): Cancellable {
        return TaskQueue.enqueue(this, callback, callbackOnUiThread)
    }

    /**
     * Performs the request and returns the response, or throws an exception if unable
     * to do so. May return `null` if this call was canceled.
     */
    protected fun call(): CallResult<To> {
        check(!ThreadUtils.runningOnUiThread()) { "This method must not be called from the main thread!" }

        val httpResult: HttpResult = callExecute()
        val callResult: CallResult<To>

        if (_cancelled.get()) {
            callResult = onCancelled()
        }
        else {
            callResult = httpResult.fold(
                onSuccess = { responseEntity ->

                    val httpStatus = responseEntity.httpStatus

                    if (httpStatus.is1xxInformational || httpStatus.is2xxSuccessful) {
                        onSucceeded(responseEntity)
                    }
                    else {
                        val cause = ResponseException(responseEntity)
                        onFailed(ApplicationLayerError(cause))
                    }
                },
                onFailure = { error ->

                    val cause = when (error) {
                        // Wrap the HTTP connection error
                        is IOException -> ConnectionException(error)
                        else -> error
                    }

                    when (cause) {
                        is CancellationException -> onCancelled()
                        else -> onFailed(TransportLayerError(cause))
                    }
                },
            )
        }

        return callResult
    }

    /**
     * TODO
     */
    protected abstract fun callExecute(): HttpResult

    /**
     * TODO
     */
    protected fun createClient(): RestApiClient {

        // Create HTTP client
        return RestApiClient.Builder()
            .httpClientConfig(httpClientConfig())
            .build()
    }

    /**
     * TODO
     */
    protected open fun httpClientConfig(): HttpClientConfig {
        return this.httpClientConfig
    }

    /**
     * TODO
     */
    protected fun createRequestEntity(): RequestEntity<HttpBody> {

        // Create HTTP request entity
        return BasicRequestEntity.Builder<HttpBody>()
            .link(link())
            .httpHeaders(httpHeaders())
            .cookieStore(cookieStore())
            .body(httpBody())
            .build()
    }

    /**
     * TODO
     */
    protected open fun link(): URI {
        return this.requestEntity.link
    }

    /**
     * TODO
     */
    protected open fun httpHeaders(): HttpHeaders {
        return HttpHeaders().also {
            it.putAll(this.requestEntity.httpHeaders)
        }
    }

    /**
     * TODO
     */
    protected open fun cookieStore(): CookieStore {
        return this.requestEntity.cookieStore
    }

    /**
     * TODO
     */
    protected open fun httpBody(): HttpBody? {
        return this.requestEntity.body
    }

    /**
     * TODO
     */
    protected abstract fun onSucceeded(responseEntity: ResponseEntity<ByteArray>): CallResult<To>

    /**
     * TODO
     */
    protected abstract fun onFailed(restApiError: RestApiError): CallResult<To>

    /**
     * TODO
     */
    protected abstract fun onCancelled(): CallResult<To>

    /**
     * TODO
     */
    final override fun clone(): Task<Ti, To> {
        return createBuilder().build()
    }

    /**
     * TODO
     */
    abstract fun createBuilder(): TaskBuilder<Ti, To>

    /**
     * TODO
     */
    final override fun cancel(): Boolean {
        return !_cancelled.getAndSet(true)
    }

// MARK: - Private Methods

    private fun yieldTo(callback: Callback<Ti, To>, callResult: CallResult<To>?) {

        if (_cancelled.get()) {
            callback.onCancelled(this)
        }
        else if (callResult != null) {
            if (callResult.isSuccess) {
                val value = checkNotNull(callResult.value()) { "value is null" }
                callback.onSucceeded(this, value)
            }
            else {
                val exception = checkNotNull(callResult.error()) { "exception is null" }
                when (exception) {
                    is CancellationException -> callback.onCancelled(this)
                    is RestApiError -> callback.onFailed(this, exception)
                    else -> error("Unsupported exception: ${exception.javaClass.name}")
                }
            }
        }
        else {
            error("!isCancelled() && (callResult == null)")
        }
    }

// MARK: - Inner Types

    abstract class Builder<Ti: HttpBody, To, BuilderType: TaskBuilder<Ti, To>>:
        TaskBuilder<Ti, To> {

        protected constructor()

        protected constructor(builder: Builder<Ti, To, BuilderType>) {
            this.tag = builder.tag
            this.requestEntity = builder.requestEntity
            this.httpClientConfig = builder.httpClientConfig
        }

        protected constructor(task: AbstractTask<Ti, To>) {
            this.tag = task.tag
            this.requestEntity = task.requestEntity
            this.httpClientConfig = task.httpClientConfig
        }

        internal var tag: String? = null
            private set

        internal var requestEntity: RequestEntity<Ti>? = null
            private set

        internal var httpClientConfig: HttpClientConfig? = null
            private set

        override fun tag(tag: String?): BuilderType {
            this.tag = tag
            return this as BuilderType
        }

        override fun requestEntity(requestEntity: RequestEntity<Ti>): BuilderType {
            this.requestEntity = requestEntity.clone()
            return this as BuilderType
        }

        override fun httpClientConfig(httpClientConfig: HttpClientConfig): BuilderType {
            this.httpClientConfig = httpClientConfig.clone()
            return this as BuilderType
        }

        abstract override fun build(): AbstractTask<Ti, To>

        abstract override fun clone(): Builder<Ti, To, BuilderType>
    }

// MARK: - Companion

    companion object {

        private fun createTag(): String {
            return "urn:tag:" + UUID.randomUUID().toBase62()
        }
    }

// MARK: - Variables

    private val _cancelled = AtomicBoolean(false)
}
